asset/asset.go

```diff
@@ -741,9 +741,38 @@ type Asset struct {
 	GroupKey *GroupKey
 }
 
+// newAssetOptions is a struct that is used to customize how a new asset is to
+// be created.
+type newAssetOptions struct {
+	assetVersion Version
+}
+
+// defaultNewAssetOptions returns the default set of asset versions.
+func defaultNewAssetOptions() *newAssetOptions {
+	return &newAssetOptions{
+		assetVersion: V0,
+	}
+}
+
+// NewAssetOpt is used to modify how a new asset is to be created.
+type NewAssetOpt func(*newAssetOptions)
+
+// WithAssetVersion can be used to create an asset with a custom version.
+func WithAssetVersion(v Version) NewAssetOpt {
+	return func(o *newAssetOptions) {
+		o.assetVersion = v
+	}
+}
+
 // New instantiates a new asset with a genesis asset witness.
 func New(genesis Genesis, amount, locktime, relativeLocktime uint64,
-	scriptKey ScriptKey, groupKey *GroupKey) (*Asset, error) {
+	scriptKey ScriptKey, groupKey *GroupKey,
+	opts ...NewAssetOpt) (*Asset, error) {
+
+	options := defaultNewAssetOptions()
+	for _, opt := range opts {
+		opt(options)
+	}
 
 	// Collectible assets can only ever be issued once.
 	if genesis.Type != Normal && amount != 1 {
```

Comment:
At this point we check that the genesis type corresponds to the `amount` field value. Maybe we should check that the asset value is valid at this point also.

Reply:
What do you mean? That the value doesn't overflow, or something else?

Reply:
I think I meant asset version (rather than "value"). I think we should check the version here also. We have an opportunity to make sure that `options.assetVersion` is sensible before even constructing the asset.